### PR TITLE
Basic Tile Placement

### PR DESCRIPTION
**Pull request by kemcbride**
OK, this is a biggie.

Things we have in here:

1. PlaceTile message type
  * takes only a MapPosition
2. Tile & Square rotation
  * shout out to stack overflow for matrix rotation tools.
3. Actual Tile Placement logic
  * how it works:
    * check if any heisters are at matching doors
    * if so, figure out what direction they'd be dropping the new tile in
    * draw the new tile (currently, the deck only contains tiles 1a and 2)
    * based on the direction of the new placement, figure out: necessary rotations, and new tile top corner position
    * add tile to game_state.tiles
4. Unit test for having Orange move one square north, and sending a PlaceTile message with the tile entrance that Orange would like to move into/place the tile on.

### CODE REVIEW

File: src/types.rs [hunk 1]
```diff
@@ -165,6 +174,58 @@ impl Tile {
...
+
+    pub fn from_matrix(m: Vec<Vec<Square>>, name: String, position: MapPosition) -> Tile {
+        let mut squares: Vec<Square> = Vec::new();
+        for row in 0..4 {
+            for col in 0..4 {
+                let square = m.get(row).unwrap().get(col).unwrap();
+                squares.push(*square);
+            }
+        }
+        info!("Size of squares: expected 16 - {}", squares.len());
```
Comment (banool) [inline src/types.rs:202]: `debug` or remove.

File: src/types.rs [hunk 2]
```diff
@@ -165,6 +174,58 @@ impl Tile {
...
+        Tile {
+            position,
+            squares,
+            name,
+        }
+    }
+
+    pub fn rotate_matrix_clockwise(m: &Vec<Vec<Square>>) -> Vec<Vec<Square>> {
```
Comment (banool) [inline src/types.rs:210]: BIIIIGGG BRAIN

File: src/game.rs [hunk 3]
```diff
@@ -25,9 +28,30 @@ pub enum MoveValidity {
...
+        // TODO: implement proper decks & base them off of GameOptions
+        let tile_deck = load_map::load_tiles_from_json();
         Game {
             game_handle,
             game_state,
+            tile_deck,
+        }
+    }
+
+    fn draw_tile(&mut self) -> Option<&Tile> {
```
[the remark this answers is not in the document]
Comment (kemcbride) [reply src/game.rs:40]: stack :o doable

File: src/game.rs [hunk 4]
```diff
@@ -233,17 +386,66 @@ impl Game {
...
 
+    fn place_tile(&mut self, position: &MapPosition, direction: &MoveDirection) -> MoveValidity {
+        let tile = self.draw_tile();
+        match tile {
+            // TODO: figure out how to handle mismatched door case for {Color}Door
+            // into TileEntrance (mismatched/asymmetric walls on tile bounds)
+            Some(t) => {
+                let new_pos = Self::new_tile_position(position, direction);
+                let num_rotations = match direction {
+                    MoveDirection::North => 1,
```
[the remark this answers is not in the document]
Comment (kemcbride) [reply src/game.rs:397]: yeah... 
..= ok

that was basically the thing. i had it 0..(n+1) and then decided, screw it, i'll just have the map return the correct value, but that is the right solution yeah

File: src/game.rs [hunk 5]
```diff
@@ -233,17 +386,66 @@ impl Game {
...
+                let rotated_tile = &Tile::from_matrix(m, t.name.clone(), new_pos);
+                self.game_state.tiles.push(rotated_tile.clone());
+
+                info!(
+                    "Added Tile {} at {:?} to Game map",
+                    rotated_tile.name, rotated_tile.position
+                );
+                MoveValidity::Valid
+            }
+            None => MoveValidity::Invalid("Couldn't draw tile from deck".to_string()),
```
[the remark this answers is not in the document]
Comment (kemcbride) [reply src/game.rs:416]: yep!

File: src/game.rs [hunk 6]
```diff
@@ -233,17 +386,66 @@ impl Game {
...
+        }
+    }
+
+    fn process_tile_placement(&mut self, pt: PlaceTile) -> MoveValidity {
+        let grid = self.get_absolute_grid();
+        let heister_to_tile_entrance_locs = self.heister_to_tile_entrance_positions(&grid);
+        let heister_pos = heister_to_tile_entrance_locs
+            .iter()
+            .find(|&(_, te)| te == &pt.tile_entrance)
+            .expect("When placing a tile, the heister must be at a tile-reveal door")
```
[the remark this answers is not in the document]
Comment (kemcbride) [reply src/game.rs:426]: Sounds legit. (y)

File: src/game.rs [hunk 7]
```diff
@@ -233,17 +386,66 @@ impl Game {
...
+            .0;
+
+        let heister_square = grid
+            .get(heister_pos)
+            .expect("Heister must be on valid square");
+        let dir = &Self::get_door_direction(heister_square).unwrap();
```
[the remark this answers is not in the document]
Comment (kemcbride) [reply src/game.rs:432]: ?? like keep it as unwrap or change line 431 to just be a plain old unwrap() - no expect?

File: src/game.rs [hunk 8]
```diff
@@ -425,4 +627,55 @@ pub mod tests {
         }
         info!("All walls line up");
     }
+
+    #[test]
+    pub fn test_tile_draw() -> () {
+        // We test with initial game state (1a), move Orange one square north,
```
[the remark this answers is not in the document]
Comment (kemcbride) [reply src/game.rs:633]: oooh. (y)

File: src/types.rs (hunk 2 above)
Comment (kemcbride) [reply src/types.rs:210]: yeah lol copy pasta from "Java Script Rotate Matrix 90 Degrees Clockwise Medium Article"

File: src/game.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (banool) [reply src/game.rs:432]: Oh I mean lines 429 -> 432 are all fine and good. In general tho, `expect` > `unwrap` bc it lets you know what was wrong.

File: src/game.rs [hunk 9]
```diff
@@ -25,9 +28,30 @@ pub enum MoveValidity {
...
+    fn draw_tile(&mut self) -> Option<&Tile> {
+        // TODO: shuffle tiles
+        match self.game_state.remaining_tiles {
+            0 => None,
+            _wildcard => {
+                let used_tile_names: HashSet<String> =
+                    HashSet::from_iter(self.game_state.tiles.iter().map(|t| t.name.clone()));
+                let all_tile_names = self.tile_deck.keys();
+                let unused_tile_names =
+                    Vec::from_iter(all_tile_names.filter(|t| !used_tile_names.contains(t.clone())));
```
[the remark this answers is not in the document]
Comment (kemcbride) [reply src/game.rs:49]: I guess the weird thing is that I'm using filter to do set uhhh intersection

File: src/game.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (kemcbride) [reply src/game.rs:426]: eep. makes sense though. sounds good.

File: src/game.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (kemcbride) [reply src/game.rs:432]: hey what do you mean here. are these expects bad? is the unwrap ok? huh?!

[the remark this answers is not in the document]
Comment (banool) [reply src/game.rs:432]: It is perfect as is, but would be perfecter if the unwrap was an expect instead!

[the remark this answers is not in the document]
Comment (kemcbride) [reply src/game.rs:432]: ok yeah makes sense. unwrap -> expect, sounds fine.

File: src/game.rs [hunk 10]
```diff
@@ -28,31 +28,18 @@ pub enum MoveValidity {
...
-                let unused_tile_names =
-                    Vec::from_iter(all_tile_names.filter(|t| !used_tile_names.contains(t.clone())));
-                let drawn_tile_name = unused_tile_names.first();
-
-                self.game_state.remaining_tiles -= 1;
-                self.tile_deck.get(drawn_tile_name.unwrap().clone())
-            }
-        }
+    fn draw_tile(&mut self) -> Option<Tile> {
+        let tile = self.tile_deck.pop();
```
Comment (banool) [inline src/game.rs:40]: NAIIICCUUUU

**Pull request merged**